### PR TITLE
Overhaul PyROS Preprocessor Subroutine and Subproblem Objects

### LINKED ISSUE #2964: Restore PyROS failed separation problem test
## Summary

#2953 marked `pyomo/contrib/pyros/tests/test_grcs.py::RegressionTest::test_discrete_separation_subsolver_error` as an expected failure in 30978f3 because invalid constants now raise an `InvalidValueError` in the problem writer.  PyROS needs to be updated to correctly catch / gracefully handle the `InvalidValueError` (and the test re-enabled).

### PR DESCRIPTION
**Pull request by shermanjasonaf**
## Summary/Motivation:

Driven by recent algorithmic developments of the PyROS preprocessor, this PR implements sweeping changes to the PyROS preprocessor subroutine and subproblem formulations/object structures.

## Changes proposed in this PR:
Update PyROS as follows:
- **Overhaul the preprocessor**
  - Leverage decision rules and equality constraints to identify second-stage variables and state variables that are mathematically nonadjustable
  - More structured methods for reformulating variable domains/bounds to constraints
  - More structured methods for rearranging inequality constraints
  - Rewrite and extend coefficient matching: carefully account for state variable-independent constraints that cannot be reformulated by casting to inequality constraint pairs
- **Update the subproblem formulations and modeling objects**
  - Modify block structures of the working model, master-like models, and separation model: partition the constraints by stage
  - Change the objective of the decision rule polishing problem to the 1-norm of the *nonstatic* terms of the decision rule expressions
  - Perform more careful initialization of the auxiliary uncertain parameter variables of the separation problem
- **Update the `UncertaintySet` class and pre-implemented subclasses** to facilitate changes to the subproblems
   - Modify return protocol of `UncertaintySet.set_as_constraint` to make auxiliary uncertain parameters easier to track
   - Add new efficient methods for calculation of auxiliary uncertain parameter values for the `CardinalitySet` and `FactorModelSet` classes 
   - Restrict range of valid `FactorModelSet` instances to those for which `psi_mat` is full column rank
   - Bug fix(es) for the `IntersectionSet`
- **Logging updates**
   - Make log of the model component statistics slightly more informative
   - Add DEBUG-level messages for logging of the individual preprocessor steps
- **Documentation updates**
   - Update "Methodology Overview" section
   - Add section on PyROS installation instructions
   - Adapt PyROS documentation to reorganized Pyomo documentation (post #3382)
- **Make tests more rigorous and extensive**
  - Address #2964
  - Use expression comparison methods (`core.expr.compare.assertExpressionsEqual`) for more precise model testing
  - Separate testing modules for the:
    - preprocessor
    - subproblem formulations
    - uncertainty set classes
      - more rigorous testing for `point_in_set`, `set_as_constraint`
    - PyROS solver as a whole
 
## TODO
- [x] (After PR #3324 merged) use `ParameterizedQuadraticRepnVisitor` in lieu of `generate_standard_repn`
- [x] Update PyROS version number and changelog

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by mrmundt — CHANGES_REQUESTED** [review]
Because this is an overhaul, please convert uses of `master` to a more inclusive word (e.g., `primary`, `main`).

**Review by jsiirola — APPROVED** [review]
Some minor things (we can either address quickly or defer to a subsequent PR), but nothing that would prevent merging.

**Comment by mrmundt** [comment]
@shermanjasonaf - One failed test on Jenkins across all Python versions:
```
pyomo.contrib.pyros.tests.test_grcs.RegressionTest.test_coefficient_matching_nonlinear_expr (from pytest) 
self = <pyomo.contrib.pyros.tests.test_grcs.RegressionTest testMethod=test_coefficient_matching_nonlinear_expr>

    @unittest.skipUnless(baron_license_is_valid, "BARON solver license is invalid.")
    def test_coefficient_matching_nonlinear_expr(self):
        """
        Test behavior of PyROS solver for model with
        equality constraint that cannot be reformulated via
        coefficient matching due to nonlinearity.
        """
        # Write the deterministic Pyomo model
        m = ConcreteModel()
        m.x1 = Var(initialize=0, bounds=(0, None))
        m.x2 = Var(initialize=0, bounds=(0, None))
        m.u = Param(initialize=1.125, mutable=True)
    
        m.con = Constraint(expr=m.u ** (0.5) * m.x1 - m.u * m.x2 <= 2)
        m.eq_con = Constraint(
            expr=m.u**2 * (m.x2 - 1)
            + m.u * (m.x1**3 + 0.5)
            - 5 * m.u * m.x1 * m.x2
            + m.u * (m.x1 + 2)
            == 0
        )
        m.obj = Objective(expr=(m.x1 - 4) ** 2 + (m.x2 - 1) ** 2)
    
        interval = BoxSet(bounds=[(0.25, 2)])
    
        # Instantiate the PyROS solver
        pyros_solver = SolverFactory("pyros")
    
        # Define subsolvers utilized in the algorithm
        local_subsolver = SolverFactory('baron')
        global_subsolver = SolverFactory("baron")
    
        # Call the PyROS solver
        with LoggingIntercept(module="pyomo.contrib.pyros", level=logging.DEBUG) as LOG:
            results = pyros_solver.solve(
                model=m,
                first_stage_variables=[m.x1],
                second_stage_variables=[m.x2],
                uncertain_params=[m.u],
                uncertainty_set=interval,
                local_solver=local_subsolver,
                global_solver=global_subsolver,
                options={
                    "objective_focus": ObjectiveType.worst_case,
                    "solve_master_globally": True,
                    "decision_rule_order": 1,
                },
            )
    
        pyros_log = LOG.getvalue()
        self.assertRegex(
            pyros_log, r".*Equality constraint '.*eq_con.*'.*cannot be written.*"
        )
    
        # should still solve in spite of coefficient matching
        # failure
>       self.assertEqual(
            results.pyros_termination_condition,
            pyrosTerminationCondition.robust_optimal,
        )
E       AssertionError: <pyrosTerminationCondition.subsolver_error: 4> != <pyrosTerminationCondition.robust_optimal: 1>

pyomo/pyomo/contrib/pyros/tests/test_grcs.py:1806: AssertionError
```

**Review by mrmundt — APPROVED** [review]

File: pyomo/contrib/pyros/tests/test_grcs.py [hunk 1]
```diff
@@ -1747,7 +1747,8 @@ def test_coefficient_matching_robust_infeasible_proof_in_pyros(self):
         )
 
     @unittest.skipUnless(
-        scip_available and scip_license_is_valid, "SCIP not available and licensed."
+        (baron_available and baron_license_is_valid and baron_version == (24, 5, 8, 0)),
+        "BARON 24.5.8 not available and licensed.",
```
Comment (jsiirola) [inline pyomo/contrib/pyros/tests/test_grcs.py:1751]: Is this test specific to BARON 24.5.8 or does it need at least 24.5.8?  I worry that while that version may not be available in any of the CI systems (I know GHA uses the current release - currently 24.10.10), and therefore will never be run.  Would `baron_version >= (24, 5, 8)` be acceptable?

Comment (shermanjasonaf) [reply pyomo/contrib/pyros/tests/test_grcs.py:1751]: On my machine, this test:

- Times out with BARON > 24.5.8
- Passes with SCIP 8.0.3 and 9.1.0 (but, based on prior discussion, appeared to trigger a time out with the SCIP distribution used in the most recent Jenkins tests)

I am considering replacing the nonlinear equality constraint of the model used in this test with a simpler nonlinear equality constraint, as the new preprocessor (correctly) derives from the current equality two nonconvex inequalities that appear to make the current problem difficult to solve. Thoughts?

Comment (jsiirola) [reply pyomo/contrib/pyros/tests/test_grcs.py:1751]: Yeah - if we can test the same behavior with a simpler problem that is always good.

**Comment by shermanjasonaf** [comment]
@jsiirola On my machine, the test `test_grcs::RegressionTest::test_discrete_separation_invalid_value_error` fails; this appears to be a consequence of #3386.

**Comment by shermanjasonaf** [comment]
That's it for updates post #3382.

**Review by mrmundt — APPROVED** [review]
I don't have anything that actually _needs_ to be changed, just some stylistic suggestions. Overall it looks good.

**Pull request merged**